### PR TITLE
WanderingCommander

### PR DESCRIPTION
**Pull request by leogrim**
@yingjieMiao @eishay @ymatsuda @stkem

### CODE REVIEW

File: kifi-backend/modules/graph/app/com/keepit/graph/wander/WanderingCommander.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    val destinations = wander(startingPoint, preferredOutputs, steps, restartProbability)
+    getRecommendations(destinations)
+  }
+
+  private def wander(startingPoint: VertexId, preferredDestinations: Set[VertexType], steps: Int, restartProbability: Double): Map[VertexId, Int] = {
+    val journal = new DestinationJournal()
+    graph.readOnly { reader =>
+      val wanderer = reader.getNewVertexReader()
+      val scout = reader.getNewVertexReader()
+      val scoutingWanderer = new ScoutingWanderer(wanderer, scout)
```
Comment (yingjieMiao) [inline kifi-backend/modules/graph/app/com/keepit/graph/wander/WanderingCommander.scala:44]: It's not immediately clear what the wander/scout metaphor mean. 
Stepping into the ScoutingWanderer code, seems that wander is associated with source and scout is associated with destination? as in `journal.onTeleportation(wanderer, scout)`

Perhaps some renaming helps

**Comment by yingjieMiao** [comment]
added a minor comment on naming. otherwise lgtm.

We probably need some tests on very toy graphs. e.g a few vertex and some very biased weights. But I understand the trouble

File: kifi-backend/modules/graph/app/com/keepit/graph/wander/WanderingCommander.scala (hunk 1 above)
Comment (leogrim) [inline kifi-backend/modules/graph/app/com/keepit/graph/wander/WanderingCommander.scala:44]: Wanderer = "a person who travels aimlessly; a traveler." => this is the main VertexReader that moves around at "random". Now a VertexReader has access to its outgoing edges (OutgoingEdgeReader), but not to the actual destinations. So when weighting an edge, you need an extra VertexReader to inspect the destination's data. 

Scout = "a soldier or other person sent out ahead of a main force so as to gather information about the enemy's position, strength, or movements."

So basically you "send" your scout VertexReader to the destination so that you can decide whether to move there or not.
